### ISSUE TITLE
Load multiple web applications/clients from within a standalone @SpringBootApplication not working

### ISSUE THREAD

**Issue by silentsnooc** [spring-projects/spring-boot#8239]
Hi!

I am currently trying to load an additional web application during the startup of a standalone `@SpringBootApplication`. I'm running into several problems but one of them is as I run the Java application from within Eclipse, I get a lot of exceptions getting thrown at me by the embedded Tomcat server (see [full output log](http://pastebin.com/PGuHUDHr)).

I am not sure why I can do about this and I'm stuck for quite a while now for such a "simple" task. I have a [related question](http://stackoverflow.com/questions/42052345/how-can-i-deploy-a-war-file-to-the-built-in-tomcat-server-of-a-spring-boot-appli) on stackoverflow if you want to take a look at it. 

I have the web application as a WAR file inside the standalone as a resource. As the application starts/loads I want to take this resource and move/copy it to the server-home directory. From there I want to call `Tomcat#addWebapp` - and like I said - I was able to make this work for the final standalone jar but I'm not able to run this in Eclipse which means I cannot debug the actual server code.

In the end it's boiling down to these few lines of code:

```java
@SpringBootApplication
public class Server {

    public static void main(String[] args) {
        SpringApplication.run(EasyModelAccessServer.class, args);
    }

    @Bean
    public EmbeddedServletContainerFactory servletContainerFactory() {

        return new TomcatEmbeddedServletContainerFactory() {

            @Override
            protected TomcatEmbeddedServletContainer getTomcatEmbeddedServletContainer(Tomcat tomcat) {

                String appHome = System.getProperty(Environment.APP_HOME);
                String targetFileName = "web-0.0.1-SNAPSHOT.war";
                InputStream resourceAsStream = getClass().getClassLoader().getResourceAsStream(targetFileName);

                File target = new File(Paths.get(appHome, targetFileName).toString());

                try {

                    java.nio.file.Files.copy(resourceAsStream, target.getAbsoluteFile().toPath(), StandardCopyOption.REPLACE_EXISTING);

                    tomcat.addWebapp("/web", target.getAbsolutePath());

                } catch (ServletException ex) {
                    throw new IllegalStateException("Failed to add webapp", ex);
                } catch (Exception e) {
                    throw new IllegalStateException("Failed to add webapp", e);
                }
                
                return super.getTomcatEmbeddedServletContainer(tomcat);
            }
        };
    }

}
```

Can anybody tell me how I can do this?

**Comment by wilkinsona** [comment]
Looking at the log your other war file is using Jasper (I assume it's got some JSPs in it) and you haven't got Jasper on the classpath. What's in your pom.xml/build.gradle?

**Comment by silentsnooc** [comment]
Hi @wilkinsona!

Well, there's not much actually - it's just a naked web client. I thought - since this is getting loaded by the built-in Tomcat - all required dependencies are going to be available at that point.

```xml
<?xml version="1.0"?>
<project
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd"
	xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<modelVersion>4.0.0</modelVersion>
	<parent>
		<groupId>org.ema.server</groupId>
		<artifactId>ema-server</artifactId>
		<version>0.0.1-SNAPSHOT</version>
	</parent>

	<artifactId>web</artifactId>
	<packaging>war</packaging>
	<name>web Maven Webapp</name>
	<url>http://maven.apache.org</url>

	<dependencies>

		<dependency>
			<groupId>junit</groupId>
			<artifactId>junit</artifactId>
			<version>3.8.1</version>
			<scope>test</scope>
		</dependency>

	</dependencies>

	<build>
		<finalName>web</finalName>
	</build>

</project>
```

There's not a lot to it, you can take a look at the web module [here](https://github.com/silentsnooc/easy-model-access/tree/master/ema-server) if you want. However, you think adding Jasper dependencies to the war file would solve this issue? That would be great actually ...

**Comment by wilkinsona** [comment]
>  However, you think adding Jasper dependencies to the war file would solve this issue? That would be great actually ...

No, not to the war, but to the Spring Boot application that's running the embedded Tomcat to which you are deploying the war.

**Comment by silentsnooc** [comment]
@wilkinsona Ok, okay I thought that his would be already included - I don't understand - why isn't that already available? Or is this due to my kind of "special" architecture that I'm currently going with? I currently have a `server`module that's just gluing together the rest-api and via a tricky maven build loads the web client into the final spring application.

Do you know which dependency I have to load in order to support the embedded Tomcat?

**Comment by wilkinsona** [comment]
`org.apache.tomcat.embed:tomcat-embed-jasper`

GitHub issues aren't really the right place for this sort of thing. It looks like you are seeking some general guidance rather than reporting a bug. If adding the Jasper dependency doesn't fix the problem, then please either provide a small sample that reproduces the problem you are seeing or ask another question on Stack Overflow (things seem to have moved on beyond the related question you linked to above).

**Comment by silentsnooc** [comment]
@wilkinsona Well, I've already tried to include this as a dependency but it does not work. The core issue is that it does not seem to be possible for a standalone to simply add a web application to its build-in Tomcat and on stackoverflow no one could give me a concrete answer to this issue or how to solve it. That is why I went on github because I thought it's either not possible or somebody here can tell me how I can do this.

I am not sure if what I do is meant to be. Usually the WAR file that gets compiled to the standalone JAR contains also the web application. In my scenario I have sourced out the web client into another Maven module and I try to include it into the standalone. I am not sure if this is supported by Spring at all actually.

**Comment by wilkinsona** [comment]
> Well, I've already tried to include this as a dependency but it does not work

Why are you using 9.0.0.M15? Spring Boot uses Tomcat 8.5. Also, "it does not work" tells me almost nothing. In what way doesn't it work? What happened?

> The core issue is that it does not seem to be possible for a standalone to simply add a web application to its build-in Tomcat

That is absolutely possible (although unusual). Your question on Stack Overflow already links to [a question and answer](http://stackoverflow.com/questions/31374726/spring-boot-how-to-add-another-war-files-to-the-embedded-tomcat/31378964) telling you how to do it. You've said in your question of Stack Overflow that "it appears that such a war file must either be located somewhere in the file system". That's correct, but your description above shows that you've already figured out how to extract the war from the fat jar and deploy it. As far as I can tell, you know the answer to your Stack Overflow question.

> In my scenario I have sourced out the web client into another Maven module and I try to include it into the standalone

Sorry, I don't understand this. What's a "web client" in this context? A war typically isn't a web client.

>  I am not sure if this is supported by Spring at all actually.

It's not really a question of what Spring (Boot) supports as it has very little to do with it. It's a question of what Tomcat supports and it definitely supports deploying a war file.

This issue isn't really making any progress. Sorry, but I don't have the time to try and guess exactly what it is that you're trying to do or what might have happened when all you say is that "it doesn't work". I'll ask one last time: please provide a sample that reproduces the problem you're facing or ask another question on Stack Overflow. If you can't or won't spend the time doing that then I can't justify spending any more time looking at this.

**Comment by silentsnooc** [comment]
@walterjs I'm sorry it was not my intention to waste anybody's time her. The problem with including the dependency into the pom.xml is that the same error messages occur than before. So nothing changes if I do that.

It is correct that I realized I would have to extract the WAR file from the standalone JAR and move it to the file system in order to load it with the build-in Tomcat. However, I actually hoped that I'd get told that this is actually not necessary because there's another way to do that.

By "web client" I was referring to the "client-only" WAR which I indent to deploy at runtime to the build-in Tomcat. Sorry if that was a confusing terminology but in my case the WAR file contains really just a simple web client.

The quickest way to show a reproduceable example would be to refer to the [repository of the server](https://github.com/silentsnooc/easy-model-access). 

The following should be enough to get it up and running:

```
mkdir ~/.ema
git clone https://github.com/silentsnooc/easy-model-access
cd easy-model-access/ema-server
mvn clean install
java -jar server/target/server-*.jar
```

The directory `~/.ema` is required at the moment. It is where the WAR is being moved as the server starts. I've tested this locally and it _should_ work but since this is under construction I cannot guarantee of course.

Anyway, I am already feeling bad so no heart feelings if you don't have time for that in which case I'll keep trying it myself until it starts to work somehow. ^^

**Comment by silentsnooc** [comment]
Well, I've notice that the [question](http://stackoverflow.com/q/31374726/826983) has an addition to the answer which went unnoticed from my side. However, even after changing the `ContextClassLoader` the result stays the same. Which thread should be the one that provides the context class loader? I surely isn't the main thread ..

**Comment by wilkinsona** [comment]
I'm not sure that relying on the thread context class loader is particularly good advice. IMO, it's better to set the parent class loader on the context explicitly:

```
tomcat.addWebapp("/web", target.getAbsolutePath()).setParentClassLoader(getClass().getClassLoader());
```

Add the call to `setParentClassLoader` and a dependency on Jasper was enough for your server to start for me.

**Comment by silentsnooc** [comment]
I don't get this.. I don't see where the difference is to my setup. If the server starts, that's one thing but do you receive something from http://localhost:8080/rest-api/dl4j/we/getAvailableModels and can you also load the web client on http://localhost:8080/web ? 

----

If I place 

	<dependency>
		<groupId>org.apache.tomcat.embed</groupId>
		<artifactId>tomcat-embed-jasper</artifactId>
		<version>8.5.0</version>
	</dependency>

in `web/pom.xml` and set the class loader as described I land on the next exception:

```
Caused by: java.lang.ClassNotFoundException: javax.servlet.jsp.tagext.TagVariableInfo
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381) ~[na:1.8.0_121]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_121]
	at org.springframework.boot.loader.LaunchedURLClassLoader.loadClass(LaunchedURLClassLoader.java:94) ~[server-0.0.1-SNAPSHOT.jar:na]
```

I can fix that by adding `jsp-api` to the `server/pom.xml` 

```
<dependency>
  <groupId>javax.servlet.jsp</groupId>
  <artifactId>jsp-api</artifactId>
  <version>2.1</version>
</dependency>
```

but after adding that and recompiling everything I'm just getting stuck at another exception:

```
Caused by: java.lang.IndexOutOfBoundsException: null
	at org.springframework.boot.loader.data.RandomAccessDataFile.getSubsection(RandomAccessDataFile.java:105) ~[server-0.0.1-SNAPSHOT.jar:na]
	at org.springframework.boot.loader.jar.JarFileEntries.getEntryData(JarFileEntries.java:212) ~[server-0.0.1-SNAPSHOT.jar:na]
	at org.springframework.boot.loader.jar.JarFileEntries.getInputStream(JarFileEntries.java:188) ~[server-0.0.1-SNAPSHOT.jar:na]
	at org.springframework.boot.loader.jar.JarFileEntries.getInputStream(JarFileEntries.java:180) ~[server-0.0.1-SNAPSHOT.jar:na]
```

I've pushed these changes so in case you want to convince yourself that I indeed applied the changes just make a `git pull` but I'm afraid you'll have to show me the way to your donation button because otherwise I'd have a bad conscience for the next week.

**Comment by wilkinsona** [comment]
This is going to be my last attempt to help you here.

You haven't added a dependency on Jasper as far as I can see. Also, there should be no need for any other dependencies: `TagVariableInfo` is in `tomcat-embed-jasper`. You should use 8.5.11 rather than 8.5.0. More generally, if you used Spring Boot's dependency management you wouldn't have to worry about the version at all.

Running with java -jar http://localhost:8080/rest-api/dl4j/we/getAvailableModels responds with:

```
HTTP/1.1 200
Content-Type: application/json;charset=UTF-8
Date: Mon, 13 Feb 2017 18:37:40 GMT
Transfer-Encoding: chunked

{
    "availableModels": {}
}
```

http://localhost:8080/web only seems to support `POST`:

```
HTTP/1.1 405
Allow: POST
Content-Length: 0
Date: Mon, 13 Feb 2017 18:39:01 GMT
```

**Comment by silentsnooc** [comment]
I've finally managed - not sure what the heck I messed up every time I started the server - and I think I might got there at least once but oversaw, or didn't think, the `Allow: POST` and at this point I guess I thought that I loaded the webapp in a wrong way or I'm still missing something and went on. So I guess now I need to find out why it's not responding to the `GET` request and delivers the static content. 

However, thank you very much for you effort here - and sorry for being a bit of an annoyance here.
